® improve <ix-drawer> accessibility, animation structure, and state handling.

priestmoon50 opened the pull request:
refactor: improve <ix-drawer> accessibility, animation structure, and state handling

## 🛠️ Drawer Component Refactor – `ix-drawer`

### Summary

This PR refactors the `<ix-drawer>` component to improve readability, accessibility, and maintainability without altering its public API or core StencilJS architecture. The changes remain fully backward-compatible and align with best practices in modern Web Component development.

---

### 🔧 `drawer.tsx` updates

- **Type Safety:** Replaced untyped `any` usage with `PointerEvent` for improved type safety and clarity.
- **Event Handling:** Migrated from `mousedown` to `pointerdown` to ensure broader device compatibility, including touch input.
- **State Refactor:** Removed the redundant `toggle` flag and introduced a clean internal state using `@State() internalVisible` to track visibility.
- **Animation Cleanup:** Merged `slideInRight` and `slideOutRight` into a unified `animateDrawer()` function to reduce repetition and improve testability.
- **Accessibility Improvements:** Added `role="dialog"` and `aria-modal="true"` to the `<Host>` element for better screen reader support.

---

### 🎨 `drawer.scss` updates

- **Layout Optimization:** Changed container display from `block` to `flex` to improve layout structure and alignment consistency.
- **Smooth Transitions:** Added CSS `transition` for `width` and `opacity` to ensure graceful UI feedback, even without JS animations.
- **Keyboard Navigation:** Introduced `:focus-visible` outline styling to improve usability for keyboard users and meet a11y standards.
- **Content Styling Polish:** Added padding and `box-sizing: border-box` to `.content` for better spacing and cleaner structure.

---

## 💡 What is the current behavior?

The drawer uses `mousedown` for outside-click handling, lacks ARIA attributes for accessibility, contains redundant state logic (`toggle`), and has split animation logic that could be unified.

GitHub Issue Number: _No associated issue_ — this is a standalone refactor PR.

---

## 🆕 What is the new behavior?

- Pointer event handling now supports both mouse and touch input.
- The drawer has clear internal state via `@State() internalVisible`.
- ARIA attributes make the component screen-reader friendly.
- DRY and unified animation method improves maintainability.
- Styling and layout are improved for keyboard and visual UX.

---

## ✅ Checklist

A pull request can only be merged if all of these conditions are met (where applicable):

- [x] 🦮 Accessibility (a11y) features were implemented
- [ ] 🗺️ Internationalization (i18n) – no hard coded strings
- [x] 📲 Responsiveness – components handle viewport changes and content overflow gracefully
- [ ] 📄 Documentation was reviewed/updated [siemens/ix-docs](https://github.com/siemens/ix-docs)
- [ ] 🧪 Unit tests were added/updated and pass (`pnpm test`)
- [ ] 📸 Visual regression tests were added/updated and pass ([Guide](https://github.com/siemens/ix/blob/main/CONTRIBUTING.md#visual-regression-testing))
- [x] 🧐 Static code analysis passes (`pnpm lint`)
- [x] 🏗️ Successful compilation (`pnpm build`, changes pushed)

---

## 👨‍💻 Help & support

Feel free to leave comments or suggestions. Happy to iterate if anything needs to be adjusted.

priestmoon50 commented:
Hi, this PR aims to improve internal structure and accessibility of the <ix-drawer> component. Happy to adjust anything if needed — thanks for reviewing!

danielleroux commented:
@priestmoon50 Thank your for your contribution. I did a first check, found some minor details. Beside that please check the linter we do not want to have that many inline comments if not needed

priestmoon50 commented:
> ## [![Quality Gate Passed](https://camo.githubusercontent.com/f7c890f34e0efe8252e17e9afd5a85aa22fc9c81b42a0ec3d4821af9f55f250d/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d7061737365642d323070782e706e67)](https://sonarcloud.io/dashboard?id=siemens_ix&pullRequest=1910) **Quality Gate passed**
> Issues ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [1 New issue](https://sonarcloud.io/project/issues?id=siemens_ix&pullRequest=1910&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/86e5c40907bf9e8933fe6bcf185f7076622a76a3bef4b6d2b5fd583959409aef/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f61636365707465642d313670782e706e67) [0 Accepted issues](https://sonarcloud.io/project/issues?id=siemens_ix&pullRequest=1910&issueStatuses=ACCEPTED)
> 
> Measures ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=siemens_ix&pullRequest=1910&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0.0% Coverage on New Code](https://sonarcloud.io/component_measures?id=siemens_ix&pullRequest=1910&metric=new_coverage&view=list) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0.0% Duplication on New Code](https://sonarcloud.io/component_measures?id=siemens_ix&pullRequest=1910&metric=new_duplicated_lines_density&view=list)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=siemens_ix&pullRequest=1910)

priestmoon50 commented:
ℹ️ Regarding the SonarQube suggestion to use `<dialog>` instead of the `role="dialog"`:

Since this component uses Shadow DOM (`shadow: true`), using the native `<dialog>` element would not work properly across all browsers and assistive technologies. Therefore, we opted to use `role="dialog"` and `aria-modal="true"` as the accessible and valid alternative, which complies with WAI-ARIA practices.

Reference: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog#usage_notes

priestmoon50 commented:
✅ All requested changes have been addressed:
- Removed `package-lock.json`
- Removed `@siemens/ix-icons` from root `package.json`
- Cleaned up inline comments
- Ran lint (fixed ESLint violations)

danielleroux commented on pnpm-lock.yaml:
Lock file update not necessary here :)

danielleroux commented on packages/core/src/components/drawer/drawer.tsx:
```diff
@@ -67,20 +67,19 @@ export class Drawer {
 
-  toggle = false;
-
   private static duration = 300;
   private callback = this.clickedOutside.bind(this);
   private divElement?: HTMLElement;
 
   @State() showContent = true;
+  // Added for unified state control
+  @State() internalVisible = false;
```
I would prefer to use the existing `toggle` variable and add there the `@State` decorator to it.

danielleroux commented on packages/core/src/components/drawer/drawer.tsx:
```diff
@@ -94,54 +93,40 @@ export class Drawer {
-        return;
-      }
-
+      if (defaultPrevented) return;
```
There is no need to remove the `if-else` style.

danielleroux commented on packages/core/src/components/drawer/drawer.tsx:
```diff
@@ -210,6 +179,8 @@ export class Drawer {
         data-testid="container"
         id="div-container"
+        role="dialog"
+        aria-modal="true"
```
```
Modal dialogs are when content is displayed and the user's interaction is limited to only that section until it is dismissed.
```
Source: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-modal

`aria-modal` is here not correct, even if the drawer is visible to elements in the background are still accessible

danielleroux commented on pnpm-lock.yaml:
Sorry to bother you again regarding the lock file but it not necessary to update dependencies with this PR.

danielleroux reviewed (commented):
Sorry for the long time between the reviews. There are still some opens to address until we can proceed here.

Thank you for your work 🎉
